wicket-core/src/main/java/org/apache/wicket/markup/head/filter/CspNonceHeaderResponse.java

```diff
@@ -0,0 +1,348 @@
+        attributes.add(HeaderItemAttribute.SCRIPT_SRC, String.valueOf(escapedUrl));
+        attributes.add(HeaderItemAttribute.TYPE, "text/javascript");
+        if (id != null) {
+            attributes.add(HeaderItemAttribute.ID, id);
+        }
+        if (defer) {
+            attributes.add(HeaderItemAttribute.SCRIPT_DEFER, "defer");
+        }
+        if (async) {
+            attributes.add(HeaderItemAttribute.SCRIPT_ASYNC, "async");
+        }
+        attributes.add(HeaderItemAttribute.CSP_NONCE, nonce);
+        if (!Strings.isEmpty(charset)) {
+            attributes.add("charset", charset);
+        }
+
+        JavaScriptUtils.writeJavaScriptUrl(response, attributes);
+    }
+
+    protected static void renderCssReferenceHeaderItem(
+            Response response, String id, String url, String media, String rel, String nonce
+    ) {
+        Args.notEmpty(url, "url");
+
+        HeaderItemAttributeMap attributes = new HeaderItemAttributeMap();
+        if (Strings.isEmpty(rel) == false)
+        {
+            attributes.add(HeaderItemAttribute.LINK_REL, rel);
+        } else {
+            attributes.add(HeaderItemAttribute.LINK_REL, "stylesheet");
+        }
+
+        attributes.add(HeaderItemAttribute.LINK_HREF, String.valueOf(Strings.escapeMarkup(url)));
+
+        if (Strings.isEmpty(media) == false)
+        {
+            attributes.add(HeaderItemAttribute.LINK_MEDIA, media.toString());
+        }
+        if (Strings.isEmpty(id) == false)
+        {
+            attributes.add(HeaderItemAttribute.ID, id);
+        }
+
+        attributes.add(HeaderItemAttribute.CSP_NONCE, nonce);
+
+        CssUtils.writeLinkUrl(response, attributes);
+
+        response.write("\n");
+    }
+
+    private static class JavaScriptContentWithNonceHeaderItem extends JavaScriptContentHeaderItem {
+
+        private String nonce;
+
+        public JavaScriptContentWithNonceHeaderItem(CharSequence javaScript, String id) {
+            super(javaScript, id, null);
+        }
+
+        @Override
+        public void render(Response response) {
+            String id = getId();
+            HeaderItemAttributeMap attributes = new HeaderItemAttributeMap();
+            attributes.add(HeaderItemAttribute.CSP_NONCE, this.nonce);
+            if (id != null) {
+                attributes.add(HeaderItemAttribute.ID, id);
+            }
+            JavaScriptUtils.writeJavaScript(response, getJavaScript(), attributes);
+        }
+
+        public JavaScriptContentWithNonceHeaderItem setNonce(String nonce) {
+            Args.notNull(nonce, "nonce");
+            this.nonce = nonce;
+            return this;
+        }
+    }
+
+    private static class JavaScriptReferenceWithNonceHeaderItem extends JavaScriptReferenceHeaderItem {
+
+        private String nonce;
+
+        public JavaScriptReferenceWithNonceHeaderItem(ResourceReference reference, PageParameters pageParameters, String id, boolean defer, String charset) {

```

Comment:
Can a JavaScriptReferenceHeaderItem really have a nonce? From what I understood, nonce is need for content-scripts only.

Reply:
Yes it can. If the the CSP policy is strict as in my example the nonce is needed for the reference like `<script src="jquery.js"></script>` and even for links such as `<link  href="my.css" />`, for some reason nonce for link is not documented on developer.mozilla.org, but that's how it practically works with `script-src 'unsafe-eval' 'nonce-%s'; style-src 'nonce-%s';`